File: Week_01/20/valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        pairs = {
            "}": "{",
            "]": "[",
            ")": "(",
        }
        stack = ["?"]

        for c in s:
            if c not in pairs:
                stack.append(c)
            elif stack.pop() != pairs[c]:
                return False
        return len(stack) == 1

    def test(self):
        for s, target in [
            ("{", False),
            ("{[", False),
            ("{}", True),
            ("{()}", True),
            ("{()))", False),
        ]:
            assert target is self.isValid(s), f"failed. s: {s}"


if __name__ == "__main__":
    Solution().test()

File: Week_01/20/valid_parentheses.go
package main

func isValid(s string) bool {
	mapping := map[byte]byte{
		')': '(',
		'}': '{',
		']': '[',
	}
	stack := make([]byte, 0)
	for i := 0; i < len(s); i++ {
		if v, ok := mapping[s[i]]; !ok {
			stack = append(stack, s[i])
		} else {
			if len(stack) == 0 || stack[len(stack)-1] != v {
				return false
			} else {
				stack = stack[:len(stack)-1]
			}
		}
	}

	return len(stack) == 0
}
